base/traits.jl

```diff
@@ -3,15 +3,26 @@
 ## numeric/object traits
 # trait for objects that have an ordering
 abstract type OrderStyle end
+
+"""
+    Ordered()
+
+Indicate that a type `T` guarantees that a method at most as specific as
+`isless(::T, ::T)` is defined and when called will not throw an exception.
+"""
 struct Ordered <: OrderStyle end
 struct Unordered <: OrderStyle end
 
 OrderStyle(instance) = OrderStyle(typeof(instance))
```

Comment:
I think it would be worth documenting this function too, as well as `Unordered`. This is what we do for `Base.IndexStyle`, `Base.IndexLinear` and `Base.IndexCartesian`.

Reply:
Agreed, but I was not sure what is the contract for `Unordered` is  (and in consequence for `OrderStyle`). I guess - as you noted @timholy should comment here.

My assumption would be:
* `Ordered` is opt-it (not mandatory)
* `Unordered` is soft opt-out (i.e. not guaranteeing that `isless` is undefined, but rather signaling that one should not rely on it)

Reply:
I'd say `Unordered` is anything that cannot be guaranteed to be `Ordered`, i.e., exactly what @bkamins said.

Reply:
OK. Should we at least recommend that types that support `isless` implement this trait, mentioning that it enables optimizations? Otherwise it's not clear what's the point of the trait.

Reply:
Sure, that's reasonable.

Reply:
@nalimilan - I have implemented your suggestions. Can you please have a look if this is OK?